src/Expander/TC.hs

```diff
@@ -149,34 +151,34 @@ notFreeInCtx var = do
 generalizeType :: Ty -> Expand (Scheme Ty)
 generalizeType ty = do
   canGeneralize <- filterM notFreeInCtx =<< metas ty
-  (body, (n, _)) <- flip runStateT (0, Map.empty) $
+  (body, (_, _, argKinds)) <- flip runStateT (0, Map.empty, []) $ do
     genTyVars canGeneralize ty
-  pure $ Scheme n body
+  pure $ Scheme argKinds body
 
   where
     genTyVars ::
       [MetaPtr] -> Ty ->
-      StateT (Natural, Map MetaPtr Natural) Expand Ty
+      StateT (Natural, Map MetaPtr Natural, [Kind]) Expand Ty
     genTyVars vars t = do
       (Ty t') <- lift $ normType t
       Ty <$> genVarsTyF vars t'
 
     genVarsTyF ::
       [MetaPtr] -> TyF Ty ->
-      StateT (Natural, Map MetaPtr Natural) Expand (TyF Ty)
+      StateT (Natural, Map MetaPtr Natural, [Kind]) Expand (TyF Ty)
     genVarsTyF vars (TyF ctor args) =
       TyF <$> genVarsCtor vars ctor
           <*> traverse (genTyVars vars) args
 
     genVarsCtor ::
       [MetaPtr] -> TypeConstructor ->
-      StateT (Natural, Map MetaPtr Natural) Expand TypeConstructor
+      StateT (Natural, Map MetaPtr Natural, [Kind]) Expand TypeConstructor
     genVarsCtor vars (TMetaVar v)
       | v `elem` vars = do
-          (i, indices) <- get
+          (i, indices, argKinds) <- get
           case Map.lookup v indices of
             Nothing -> do
-              put (i + 1, Map.insert v i indices)
+              put (i + 1, Map.insert v i indices, argKinds ++ [KStar])
```

Comment:
I could get rid of the `Natural`, it can be recovered from the length of the list.

Reply:
I guess I should use a `Seq` instead of a list in order to avoid quadratic performance here...

Reply:
I suspect that the list just won't be a problem here. A quadradic overhead on a list of at most 10 elements is still tiny.